File: backend/src/application/use-cases/CreateCommunity.ts
import { Community } from '../../domain/entities/Community';
import { ICommunityRepository } from '../../domain/repositories/ICommunityRepository';

export interface CreateCommunityDTO {
  name: string;
  description: string;
  latitude: number;
  longitude: number;
  radiusKm: number;
  country: string;
  region: string;
}

export class CreateCommunity {
  constructor(private readonly communityRepository: ICommunityRepository) {}

  async execute(dto: CreateCommunityDTO): Promise<Community> {
    const existing = await this.communityRepository.findByName(dto.name);
    if (existing) {
      throw new Error('A community with this name already exists');
    }

    const community = new Community({
      ...dto,
      isActive: true,
    });

    return this.communityRepository.save(community);
  }
}

File: backend-dotnet/src/Hestia.Application/UseCases/CreateCommunity.cs
using Hestia.Application.DTOs;
using Hestia.Domain.Entities;
using Hestia.Domain.Enums;
using Hestia.Domain.Repositories;

namespace Hestia.Application.UseCases;

public class CreateCommunity
{
    private readonly ICommunityRepository _communityRepository;
    private readonly IUserRepository _userRepository;

    public CreateCommunity(ICommunityRepository communityRepository, IUserRepository userRepository)
    {
        _communityRepository = communityRepository;
        _userRepository = userRepository;
    }

    public async Task<Community> ExecuteAsync(CreateCommunityRequest dto, Guid creatorId, CancellationToken ct = default)
    {
        var existing = await _communityRepository.FindByNameAsync(dto.Name, ct);
        if (existing is not null)
            throw new ApplicationException("A community with this name already exists");

        var community = new Community(
            dto.Name, dto.Description,
            dto.Latitude, dto.Longitude,
            dto.RadiusKm, dto.Country, dto.Region);

        var saved = await _communityRepository.SaveAsync(community, ct);

        // Auto-join creator as coordinator
        var creator = await _userRepository.FindByIdAsync(creatorId, ct);
        if (creator is not null)
        {
            creator.JoinCommunity(saved.Id);
            creator.Role = UserRole.Coordinator;
            await _userRepository.UpdateAsync(creator, ct);
            saved.IncrementMemberCount();
            await _communityRepository.UpdateAsync(saved, ct);
        }

        return saved;
    }
}
